### PR TITLE
Adds bullets to the game simulation.

### PR DESCRIPTION
**Pull request by EthanSteinberg**
In addition, this refactors the client code so that the game state is isolated from the rest of the logic.

### CODE REVIEW

**Comment by nabijaczleweli** [comment]
> src/gamestate.rs:71:13: 71:26 warning: unused variable: `destroyed`, #[warn(unused_variables)] on by default
> src/gamestate.rs:71         let mut destroyed: Vec<(u32, u32)> = Vec::new();
>                                 ^~~~~~~~~~~~~
> src/gamestate.rs:71:13: 71:26 warning: variable does not need to be mutable, #[warn(unused_mut)] on by default
> src/gamestate.rs:71         let mut destroyed: Vec<(u32, u32)> = Vec::new();
>                                 ^~~~~~~~~~~~~

**Comment by EthanSteinberg** [comment]
Lol, oops. Fixed now.

File: server/src/gamestate.rs [hunk 1]
```diff
@@ -64,9 +109,23 @@ impl GameState {
     fn process_websocket_event(&mut self, message: WebSocketEvent) {
         match message {
             WebSocketEvent::ClientCreated { client } => {
-                let _ = self.players
-                            .insert(client.id, message::Player::not_moving(client.id, 0.0, 0.0));
-                let _ = self.clients.insert(client.id, client);
+                let welcome_message = message::Message::Welcome {
+                    id: client.id.clone(),
```
Comment (nabijaczleweli) [inline server/src/gamestate.rs]: Stop. Cloning. Primitives.

File: server/src/gamestate.rs [hunk 2]
```diff
@@ -79,13 +138,16 @@ impl GameState {
 
     /// Serialize the entire game state into one json string.
     fn serialize(&self) -> String {
-        let players = Vec::from_iter(self.players
-                                         .values()
-                                         .map(Clone::clone));
+        let players_clone = Vec::from_iter(self.players
```
Comment (nabijaczleweli) [inline server/src/gamestate.rs]: Why not just `players`?

File: client/input.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+(function() {
+	var Game = window.Game = window.Game || {};
```
Comment (nabijaczleweli) [inline client/input.js]: Tab-indent

**Comment by EthanSteinberg** [comment]
`players` is a `HashMap`. `players_clone` is a `Vec`. I guess I should fix the naming.

**Comment by nabijaczleweli** [comment]
:+1: from me for the server-side.
Not sure about client-side.

File: client/gamescreen.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
+(function() {
+    'use strict';
+
+    var Game = window.Game = window.Game || {};
+
+    //
+    // The main state holding class for the game.
```
Comment (thecoshman) [inline client/gamescreen.js:7]: Shouldn't the game state and the game rendering be separate classes?

File: client/gamescreen.js [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                move_y += 1;
+            }
+            if (keymap[Game.INPUT_MOVE_LEFT]) {
+                move_x += -1;
+            }
+            if (keymap[Game.INPUT_MOVE_RIGHT]) {
+                move_x += 1;
+            }
+
+            socket.startMoving(Game.makeVector(move_x, move_y));
```
Comment (thecoshman) [inline client/gamescreen.js]: Should normalise it really.

Also, if the vector is `(0,0)` then not only doe sit look like a funny face, but we should send a stop moving message instead.

File: client/gamescreen.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+            if (player == null) {
+                return;
+            }
+
+            var dx = x - player.position.x;
+            var dy = y - player.position.y;
+
+
+            socket.fire(Game.makeVector(dx, dy));
```
Comment (thecoshman) [inline client/gamescreen.js:65]: Normalisation? I forget the message spec

File: server/src/gamestate.rs [hunk 7]
```diff
@@ -43,10 +50,49 @@ impl GameState {
...
+            }
+        }
+
+        // Check for collisions
+        for (_, bullet) in &mut self.bullets {
+            for (_, player) in &mut self.players {
+                let dx = bullet.x - player.x;
+                let dy = bullet.y - player.y;
+                let dist = (dx * dx + dy * dy).sqrt();
+                if dist < BULLET_RADIUS + PLAYER_RADIUS {
```
Comment (thecoshman) [inline server/src/gamestate.rs:78]: Player is a square, but you are simulating it as a circle. Also, micro-optimisation time, it's better to compare the delta sqaured, rather then calculating the sqrt. We should also look to move a lot of the logic here (in `process_game_update`) into small functions with clearer intent

File: client/gamescreen.js (hunk 6 above)
Comment (nabijaczleweli) [inline client/gamescreen.js:65]: > aiming direction vector SHOULD be normalised, but the server MUST NOT assume that it is

Comment (thecoshman) [inline client/gamescreen.js:65]: Have you a good reason to not normalise?

Comment (nabijaczleweli) [inline client/gamescreen.js:65]: No

Comment (EthanSteinberg) [inline client/gamescreen.js:65]: Uh, socket.fire does the normalization.

File: client/gamescreen.js (hunk 5 above)
Comment (EthanSteinberg) [inline client/gamescreen.js]: As below, socket.fire does the normalization.
Why do we need a stop moving message? What does it do differently than the normal move message?

File: client/gamescreen.js (hunk 4 above)
Comment (EthanSteinberg) [inline client/gamescreen.js:7]: We can make a separate rendering class later if necessary, but this is the cleanest, simplest way to do it for now.

File: server/src/gamestate.rs (hunk 7 above)
Comment (EthanSteinberg) [inline server/src/gamestate.rs:78]: Do we want to make the players be squares? For the MVP, circles are probably better. They are much, much easier to simulate.

Comment (thecoshman) [inline server/src/gamestate.rs:78]: I would agree with that, but I think I have heard plenty of other talk the they should be square. Make them circles, it's easier as you say, and size is a good enough distinction.

File: client/gamescreen.js (hunk 6 above)
Comment (thecoshman) [inline client/gamescreen.js:65]: oh, well as long as it is done before sending, then I'm happy enough :smile:

File: client/gamescreen.js (hunk 5 above)
Comment (thecoshman) [inline client/gamescreen.js]: If we don't send a stop moving message, server should presume the player is still moving in the direction they last said.

Comment (nabijaczleweli) [inline client/gamescreen.js]: Done

**Comment by nabijaczleweli** [comment]
Full :+1: from me
@Lalaland, please rebase-squash

**Comment by thecoshman** [comment]
:+1:

**Pull request merged**